How to create anchor for div, tables, etc.?

We can create link from header, but in my case, I have use intensive linking in page using paragraphs, images, table. Can you create some plugin for creation anchor?

```
<div id="some-unique-id" ...
<img id="some-unique-id" ...
<table id="some-unique-id" ...
```

I installed `graphviz`. Command `dot` in my PATH, test work fine:

```bash
$ sudo apt install graphviz
$ which dot
/usr/bin/dot
$ echo 'digraph { a -> b }' | dot -Tsvg > output.svg
```

Maven test error output:

```
...
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running org.testingisdocumenting.znai.utils.StringUtilsTest
. [value] equals "
        statement1;
        statement2"
    [value]:   actual: <java.lang.String>
             _______________
             
                 statement1;
                 statement2
             _______________
             
             expected: <java.lang.String>
             _______________
             
                 statement1;
                 statement2
             _______________ (58ms)
. [value] equals ""
    [value]:   actual: "" <java.lang.String>
             expected: "" <java.lang.String> (0ms)
. [value] equals 8
    [value]:   actual: 8 <java.lang.Integer>
             expected: 8 <java.lang.Integer> (1ms)
. [value] equals true
    [value]:   actual: true <java.lang.Boolean>
             expected: true <java.lang.Boolean> (1ms)
X failed expecting [value] to equal true: 
    mismatches:
    
    [value]:   actual: false <java.lang.Boolean>
             expected: true <java.lang.Boolean> (4ms)
. [value] equals "a prefix:line1 line1
             line2
             line #3"
    [value]:   actual: <java.lang.String>
             ____________________
             a prefix:line1 line1
                      line2
                      line #3
             ____________________
             
             expected: <java.lang.String>
             ____________________
             a prefix:line1 line1
                      line2
                      line #3
             ____________________ (1ms)
. [value] equals "int a = 2;
    int b = 3;"
    [value]:   actual: <java.lang.String>
             __________
             int a = 2;
             int b = 3;
             __________
             
             expected: <java.lang.String>
             __________
             int a = 2;
             int b = 3;
             __________ (1ms)
. [value] equals ""
    [value]:   actual: "" <java.lang.String>
             expected: "" <java.lang.String> (1ms)
. [value] equals "20"
    [value]:   actual: "20" <java.lang.String>
             expected: "20" <java.lang.String> (0ms)
. [value] equals "hello"
    [value]:   actual: "hello" <java.lang.String>
             expected: "hello" <java.lang.String> (1ms)
. [value] equals ""
    [value]:   actual: "" <java.lang.String>
             expected: "" <java.lang.String> (0ms)
. [value] equals ""
    [value]:   actual: "" <java.lang.String>
             expected: "" <java.lang.String> (0ms)
. [value] equals "hello world"
    [value]:   actual: "hello world" <java.lang.String>
             expected: "hello world" <java.lang.String> (1ms)
. [value] equals "hello world"
    [value]:   actual: "hello world" <java.lang.String>
             expected: "hello world" <java.lang.String> (0ms)
. [value] equals """"
    [value]:   actual: """" <java.lang.String>
             expected: """" <java.lang.String> (0ms)
. [value] equals ""hello world""
    [value]:   actual: ""hello world"" <java.lang.String>
             expected: ""hello world"" <java.lang.String> (1ms)
. [value] equals "hello "
    [value]:   actual: "hello " <java.lang.String>
             expected: "hello " <java.lang.String> (1ms)
Tests run: 9, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.965 sec <<< FAILURE!
is number(org.testingisdocumenting.znai.utils.StringUtilsTest)  Time elapsed: 0.031 sec  <<< FAILURE!
java.lang.AssertionError: 
mismatches:

[value]:   actual: false <java.lang.Boolean>
         expected: true <java.lang.Boolean>
	at org.testingisdocumenting.webtau.expectation.ActualValue.handleMismatch(ActualValue.java:135)
	at org.testingisdocumenting.webtau.expectation.ActualValue.shouldStep(ActualValue.java:89)
	at org.testingisdocumenting.webtau.expectation.ActualValue.lambda$should$0(ActualValue.java:55)
	at org.testingisdocumenting.webtau.utils.FunctionUtils.lambda$toSupplier$0(FunctionUtils.java:26)
	at org.testingisdocumenting.webtau.utils.FunctionUtils.lambda$toFunction$3(FunctionUtils.java:46)
	at org.testingisdocumenting.webtau.reporter.WebTauStep.executeSingleRunWithAction(WebTauStep.java:302)
	at org.testingisdocumenting.webtau.reporter.WebTauStep.executeSingleRun(WebTauStep.java:290)
	at org.testingisdocumenting.webtau.reporter.WebTauStep.execute(WebTauStep.java:283)
	at org.testingisdocumenting.webtau.expectation.ActualValue.executeStep(ActualValue.java:178)
	at org.testingisdocumenting.webtau.expectation.ActualValue.should(ActualValue.java:53)
	at org.testingisdocumenting.webtau.expectation.ActualValueExpectations$should.call(Unknown Source)
	at org.testingisdocumenting.webtau.data.expectation.ExpectationExtension.should(ExpectationExtension.groovy:27)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.runtime.metaclass.ReflectionMetaMethod.invoke(ReflectionMetaMethod.java:54)
	at org.codehaus.groovy.runtime.metaclass.NewInstanceMetaMethod.invoke(NewInstanceMetaMethod.java:54)
	at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite$PojoMetaMethodSiteNoUnwrapNoCoerce.invoke(PojoMetaMethodSite.java:247)
	at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite.call(PojoMetaMethodSite.java:56)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:139)
	at org.testingisdocumenting.znai.utils.StringUtilsTest.is number(StringUtilsTest.groovy:89)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)
...
```

I'm setting java in terminal like this (use `8.0.342-amzn` in current terminal):

```bash
$ sdk use java 8.0.342-amzn
```

--- a/znai-core/src/main/java/org/testingisdocumenting/znai/extensions/file/AnchorFeature.java
+++ b/znai-core/src/main/java/org/testingisdocumenting/znai/extensions/file/AnchorFeature.java
@@ -21,9 +21,12 @@ import org.testingisdocumenting.znai.extensions.PluginParams;
 import org.testingisdocumenting.znai.extensions.PluginParamsDefinition;
 import org.testingisdocumenting.znai.extensions.features.PluginFeature;
 import org.testingisdocumenting.znai.structure.DocStructure;
+import org.testingisdocumenting.znai.utils.NameUtils;
 
 import java.nio.file.Path;
 import java.util.Map;
+
+import static org.testingisdocumenting.znai.extensions.PluginParamsDefinitionCommon.TITLE_KEY;
 
 public class AnchorFeature implements PluginFeature {
     public static final PluginParamsDefinition paramsDefinition = createParamsDefinition();
@@ -43,11 +46,19 @@ public class AnchorFeature implements PluginFeature {
     @Override
     public void updateProps(Map<String, Object> props) {
         String anchorId = pluginParams.getOpts().get(ANCHOR_ID_KEY, "");
-        if (anchorId.isEmpty()) {
+        if (!anchorId.isEmpty()) {
+            docStructure.registerLocalAnchor(markupPath, anchorId);
             return;
         }
 
-        docStructure.registerLocalAnchor(markupPath, anchorId);
+        String title = pluginParams.getOpts().get(TITLE_KEY, props.getOrDefault(TITLE_KEY, "").toString());
+        if (!title.isEmpty()) {
+            String idFromTitle = NameUtils.idFromTitle(title);
+            String uniqueAnchorId = docStructure.generateUniqueAnchor(markupPath, idFromTitle);
+            docStructure.registerLocalAnchor(markupPath, idFromTitle);
+
+            props.put(ANCHOR_ID_KEY, uniqueAnchorId);
+        }
     }
 
     private static PluginParamsDefinition createParamsDefinition() {

--- a/znai-core/src/main/java/org/testingisdocumenting/znai/parser/docelement/DocElementCreationParserHandler.java
+++ b/znai-core/src/main/java/org/testingisdocumenting/znai/parser/docelement/DocElementCreationParserHandler.java
@@ -56,7 +56,6 @@ public class DocElementCreationParserHandler implements ParserHandler {
     private final Deque<DocElement> elementsStack;
 
     private String currentSectionTitle;
-    private SubHeadingUniqueIdGenerator subHeadingUniqueIdGenerator;
 
     private boolean isSectionStarted;
 
@@ -74,7 +73,6 @@ public class DocElementCreationParserHandler implements ParserHandler {
 
         this.currentSectionTitle = "";
         this.isSectionStarted = false;
-        this.subHeadingUniqueIdGenerator = new SubHeadingUniqueIdGenerator("");
     }
 
     public DocElement getDocElement() {
@@ -109,9 +107,10 @@ public class DocElementCreationParserHandler implements ParserHandler {
         props.put("title", title);
 
         start(DocElementType.SECTION, props);
-        subHeadingUniqueIdGenerator = new SubHeadingUniqueIdGenerator(id);
 
-        componentsRegistry.docStructure().registerLocalAnchor(path, id);
+        DocStructure docStructure = componentsRegistry.docStructure();
+        docStructure.registerLocalAnchor(path, id);
+        docStructure.onSectionOrSubHeading(path, 1, id);
 
         isSectionStarted = true;
     }
@@ -136,16 +135,17 @@ public class DocElementCreationParserHandler implements ParserHandler {
 
         String idByTitle = new PageSectionIdTitle(title, headingPropsMap).getId();
 
-        subHeadingUniqueIdGenerator.registerSubHeading(level, idByTitle);
-        String id = subHeadingUniqueIdGenerator.generateId();
+        DocStructure docStructure = componentsRegistry.docStructure();
+        docStructure.onSectionOrSubHeading(path, level, idByTitle);
+
+        String id = docStructure.generateUniqueAnchor(path, "");
+        docStructure.registerLocalAnchor(path, id);
 
         Map<String, Object> props = new LinkedHashMap<>(headingPropsMap);
         props.put("id", id);
         props.put("level", level);
         props.put("title", title);
-
         append(DocElementType.SUB_HEADING, props);
-        componentsRegistry.docStructure().registerLocalAnchor(path, id);
     }
 
     @Override

--- a/znai-core/src/main/java/org/testingisdocumenting/znai/structure/DocStructure.java
+++ b/znai-core/src/main/java/org/testingisdocumenting/znai/structure/DocStructure.java
@@ -17,13 +17,15 @@
 package org.testingisdocumenting.znai.structure;
 
 import java.nio.file.Path;
-import java.util.Map;
 import java.util.Optional;
 
 public interface DocStructure {
     void validateUrl(Path path, String additionalClue, DocUrl docUrl);
     String createUrl(Path path, DocUrl docUrl);
     String fullUrl(String relativeUrl);
+
+    void onSectionOrSubHeading(Path path, int level, String id);
+    String generateUniqueAnchor(Path path, String localId);
 
     void registerGlobalAnchor(Path sourcePath, String anchorId);
     void registerLocalAnchor(Path path, String anchorId);

--- a/znai-core/src/main/java/org/testingisdocumenting/znai/structure/UniqueAnchorIdGenerator.java
+++ b/znai-core/src/main/java/org/testingisdocumenting/znai/structure/UniqueAnchorIdGenerator.java
@@ -14,44 +14,62 @@
  * limitations under the License.
  */
 
-package org.testingisdocumenting.znai.parser.docelement;
+package org.testingisdocumenting.znai.structure;
 
+import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.stream.Collectors;
 
-class SubHeadingUniqueIdGenerator {
-    private final List<LevelId> subheadings;
-    private final Map<String, Integer> usedSubheadingsCount;
+public class UniqueAnchorIdGenerator {
+    private final List<LevelId> headings;
+    private final Map<String, Integer> usedHeadingsCount;
+    private Path previousPath;
 
-    SubHeadingUniqueIdGenerator(String topPrefix) {
-        subheadings = new ArrayList<>();
-        subheadings.add(new LevelId(1, topPrefix));
-
-        usedSubheadingsCount = new HashMap<>();
+    public UniqueAnchorIdGenerator() {
+        headings = new ArrayList<>();
+        usedHeadingsCount = new HashMap<>();
     }
 
-    void registerSubHeading(Integer level, String id) {
-        subheadings.removeIf(levelId -> levelId.level >= level);
-        subheadings.add(new LevelId(level, id));
+    public void registerSectionOrSubHeading(Path path, Integer level, String id) {
+        clearHeadingsOnPathChange(path);
+        headings.removeIf(levelId -> levelId.level >= level);
+
+        previousPath = path;
+        headings.add(new LevelId(level, id));
     }
 
-    String generateId() {
-        String prefix = subheadings.stream()
+    public String generateId(Path path, String nonUniqueId) {
+        clearHeadingsOnPathChange(path);
+
+        String prefix = headings.stream()
                 .map(LevelId::getId)
                 .filter(id -> !id.isEmpty())
                 .collect(Collectors.joining("-"));
 
-        Integer count = usedSubheadingsCount.get(prefix);
+        if (!nonUniqueId.isEmpty()) {
+            prefix = prefix + (prefix.isEmpty() ? nonUniqueId : "-" + nonUniqueId);
+        }
+
+        Integer count = usedHeadingsCount.get(prefix);
         if (count == null) {
-            usedSubheadingsCount.put(prefix, 1);
+            usedHeadingsCount.put(prefix, 1);
             return prefix;
         }
 
-        usedSubheadingsCount.put(prefix, count + 1);
+        usedHeadingsCount.put(prefix, count + 1);
         return prefix + "-" + (count + 1);
+    }
+
+    private void clearHeadingsOnPathChange(Path path) {
+        if (!path.equals(previousPath)) {
+            headings.clear();
+            usedHeadingsCount.clear();
+        }
+
+        previousPath = path;
     }
 
     static class LevelId {

--- a/znai-core/src/main/java/org/testingisdocumenting/znai/utils/NameUtils.java
+++ b/znai-core/src/main/java/org/testingisdocumenting/znai/utils/NameUtils.java
@@ -1,4 +1,5 @@
 /*
+ * Copyright 2022 znai maintainers
  * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
@@ -24,8 +25,8 @@ public class NameUtils {
         if (title == null)
             return null;
 
-        String onlyTextAndNumbers = title.replaceAll("[^a-zA-Z0-9-_ ]", "");
-        return onlyTextAndNumbers.toLowerCase().replaceAll("\\s+", "-");
+        String onlyTextAndNumbers = title.replaceAll("[^a-zA-Z0-9-_./ ]", "");
+        return onlyTextAndNumbers.toLowerCase().replaceAll("[\\s./]+", "-");
     }
 
     public static String dashToCamelCaseWithSpaces(final String dashBasedName) {

--- a/znai-core/src/test/groovy/org/testingisdocumenting/znai/parser/MarkdownParserTest.groovy
+++ b/znai-core/src/test/groovy/org/testingisdocumenting/znai/parser/MarkdownParserTest.groovy
@@ -20,6 +20,7 @@ package org.testingisdocumenting.znai.parser
 import org.testingisdocumenting.znai.parser.commonmark.MarkdownParser
 import org.junit.Test
 
+import java.nio.file.Path
 import java.nio.file.Paths
 
 import static org.testingisdocumenting.webtau.Matchers.code
@@ -219,7 +220,7 @@ world""")
 
     @Test
     void "second level section"() {
-        parse("## Secondary Section \ntext text")
+        parse("## Secondary Section \ntext text", Paths.get("new-file.md"))
         content.should == [[type: 'SubHeading', level: 2, title: 'Secondary Section', id: 'secondary-section'],
                            [type: 'Paragraph', content: [[type: 'SimpleText', text: 'text text']]]]
     }
@@ -514,8 +515,8 @@ title: custom title
         parseResult.pageMeta.toMap().title.should == ["custom title"]
     }
 
-    private void parse(String markdown) {
-        parseResult = parser.parse(Paths.get("test.md"), markdown)
+    private void parse(String markdown, Path path = Paths.get("test.md")) {
+        parseResult = parser.parse(path, markdown)
         content = parseResult.docElement.getContent().collect { it.toMap() }
     }
 }

--- a/znai-core/src/test/groovy/org/testingisdocumenting/znai/parser/TestDocStructure.groovy
+++ b/znai-core/src/test/groovy/org/testingisdocumenting/znai/parser/TestDocStructure.groovy
@@ -19,14 +19,15 @@ package org.testingisdocumenting.znai.parser
 
 import org.testingisdocumenting.znai.structure.DocStructure
 import org.testingisdocumenting.znai.structure.DocUrl
-import org.testingisdocumenting.znai.structure.GlobalAnchor
 import org.testingisdocumenting.znai.structure.TableOfContents
+import org.testingisdocumenting.znai.structure.UniqueAnchorIdGenerator
 
 import java.nio.file.Path
 
 class TestDocStructure implements DocStructure {
     private Set<String> validLinks = [] as Set
     private TableOfContents toc = new TableOfContents()
+    private final UniqueAnchorIdGenerator uniqueAnchorIdGenerator = new UniqueAnchorIdGenerator()
 
     Set<String> registeredLocalLinks = [] as TreeSet
 
@@ -59,6 +60,16 @@ class TestDocStructure implements DocStructure {
     @Override
     String fullUrl(String relativeUrl) {
         return  "/test-doc/" + relativeUrl
+    }
+
+    @Override
+    void onSectionOrSubHeading(Path path, int level, String id) {
+        uniqueAnchorIdGenerator.registerSectionOrSubHeading(path, level, id)
+    }
+
+    @Override
+    String generateUniqueAnchor(Path path, String localId) {
+        return uniqueAnchorIdGenerator.generateId(path, localId)
     }
 
     @Override

--- a/znai-core/src/test/groovy/org/testingisdocumenting/znai/structure/UniqueAnchorIdGeneratorTest.groovy
+++ b/znai-core/src/test/groovy/org/testingisdocumenting/znai/structure/UniqueAnchorIdGeneratorTest.groovy
@@ -1,0 +1,54 @@
+/*
+ * Copyright 2022 znai maintainers
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.testingisdocumenting.znai.structure
+
+import org.junit.Test
+
+import java.nio.file.Paths
+
+class UniqueAnchorIdGeneratorTest {
+    @Test
+    void "same anchor id"() {
+        def path = Paths.get("")
+
+        def generator = new UniqueAnchorIdGenerator()
+        generator.generateId(path, "my-image").should == "my-image"
+
+        generator.registerSectionOrSubHeading(path, 1, "section")
+        generator.registerSectionOrSubHeading(path, 2, "sub-section")
+        generator.generateId(path, "my-image").should == "section-sub-section-my-image"
+        generator.generateId(path, "my-image").should == "section-sub-section-my-image-2"
+
+        generator.registerSectionOrSubHeading(path, 2, "another-sub-section")
+        generator.generateId(path, "my-image").should == "section-another-sub-section-my-image"
+
+        generator.registerSectionOrSubHeading(path, 1, "another-section")
+        generator.generateId(path, "my-image").should == "another-section-my-image"
+    }
+
+    @Test
+    void "different paths"() {
+        def generator = new UniqueAnchorIdGenerator()
+
+        def path = Paths.get("")
+        generator.registerSectionOrSubHeading(path, 1, "section")
+        generator.registerSectionOrSubHeading(path, 2, "sub-section")
+
+        def anotherPath = Paths.get("my-file.md")
+        generator.generateId(anotherPath, "my-image").should == "my-image"
+    }
+}

--- a/znai-core/src/test/groovy/org/testingisdocumenting/znai/utils/NameUtilsTest.groovy
+++ b/znai-core/src/test/groovy/org/testingisdocumenting/znai/utils/NameUtilsTest.groovy
@@ -28,7 +28,7 @@ class NameUtilsTest {
         assert idFromTitle('') == ''
         assert idFromTitle('word') == 'word'
         assert idFromTitle('Hello DearWorld') == 'hello-dearworld'
-        assert idFromTitle('Hello!. %#$ DearWorld?') == 'hello-dearworld'
+        assert idFromTitle('Hello!. %#$ Dear/World?') == 'hello-dear-world'
         assert idFromTitle('Negative-Value Tests') == 'negative-value-tests'
     }
 

--- a/znai-docs/znai/layout/tables.md
+++ b/znai-docs/znai/layout/tables.md
@@ -37,7 +37,9 @@ Use `title` parameter to set a table title.
 
 # Anchor
 
-Use `anchorId` parameter to set table anchor for linking. Hover over title to reveal anchor link to click.
+When you specify a title, hover mouse over it to see a clickable anchor.
+
+Use `anchorId` to override auto generated identifier.
 
     :include-table: table/table.json {title: "Monthly Report", anchorId: "my-super-table"}
 

--- a/znai-docs/znai/release-notes/1.51/doc-2021-06-26-brew.md
+++ b/znai-docs/znai/release-notes/1.51/doc-2021-06-26-brew.md
@@ -1,1 +1,1 @@
-* Doc: [brew install instruction](introduction/getting-started#brew)
+* Doc: [brew install instruction](introduction/getting-started#command-line-brew)

--- a/znai-docs/znai/release-notes/1.58/add-2022-07-08-python-method-def.md
+++ b/znai-docs/znai/release-notes/1.58/add-2022-07-08-python-method-def.md
@@ -1,1 +1,1 @@
-* Add [include-python-method: package.method_name](python/auto-reference#methodfunction-definition) to display a function signature, documentation and parameters
+* Add [include-python-method: package.method_name](python/auto-reference#method-function-definition) to display a function signature, documentation and parameters

--- a/znai-docs/znai/release-notes/1.58/add-2022-07-12-python-class-properties.md
+++ b/znai-docs/znai/release-notes/1.58/add-2022-07-12-python-class-properties.md
@@ -1,1 +1,1 @@
-* Render [Python Properties](python/auto-reference#finmoneymoney-properties) as [API Parameters](snippets/api-parameters)
+* Render [Python Properties](python/auto-reference#fin-money-money-properties) as [API Parameters](snippets/api-parameters)

--- a/znai-docs/znai/release-notes/1.60/add-2022-09-09-auto-anchor-from-title.md
+++ b/znai-docs/znai/release-notes/1.60/add-2022-09-09-auto-anchor-from-title.md
@@ -1,0 +1,3 @@
+* Add: Auto anchor id from title for [Tables](layout/tables#title)
+* Add: Auto anchor id from title for [Snippets](snippets/external-code-snippets#title)
+* Add: Auto anchor id from title for [Images](visuals/images#title)

--- a/znai-docs/znai/snippets/external-code-snippets.md
+++ b/znai-docs/znai/snippets/external-code-snippets.md
@@ -51,14 +51,13 @@ Use the `autoTitle` property to set `title` to be the file name.
 
 # Anchor
 
-Use `anchorId` to make a code snippet linkable. Hover mouse over title to see a clickable anchor.
+When you specify a title, hover mouse over it to see a clickable anchor.
+
+Use `anchorId` to override auto generated identifier. 
 
     :include-file: file-name.js {autoTitle: true, anchorId: "my-code-anchor"}
 
 :include-file: file-name.js {autoTitle: true, anchorId: "my-code-anchor"}
-
-
-Note: title is required for anchor to work
 
 # Wide Code
 

--- a/znai-docs/znai/visuals/images.md
+++ b/znai-docs/znai/visuals/images.md
@@ -34,7 +34,9 @@ Use `title` to add a title to an image.
 
 # Anchor
 
-Use `anchorId` to add an anchor to the image. Hover over title to reveal anchor link to click.
+When you specify a title, hover mouse over it to see a clickable anchor.
+
+Use `anchorId` to override auto generated identifier.
 
     :include-image: castle.jpg {title: "beautiful castle", anchorId: "castle-image"}
 

--- a/znai-website-gen/src/main/java/org/testingisdocumenting/znai/website/WebSiteDocStructure.java
+++ b/znai-website-gen/src/main/java/org/testingisdocumenting/znai/website/WebSiteDocStructure.java
@@ -27,7 +27,6 @@ import java.net.HttpURLConnection;
 import java.net.URL;
 import java.nio.file.Path;
 import java.util.*;
-import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 import static java.util.stream.Collectors.joining;
@@ -43,6 +42,8 @@ class WebSiteDocStructure implements DocStructure {
     private final Map<String, GlobalAnchor> globalAnchorsById;
     private final Map<TocItem, List<String>> localAnchorIdsByTocItem;
 
+    private final UniqueAnchorIdGenerator uniqueAnchorIdGenerator;
+
     WebSiteDocStructure(ComponentsRegistry componentsRegistry,
                         DocMeta docMeta,
                         TableOfContents toc,
@@ -55,6 +56,7 @@ class WebSiteDocStructure implements DocStructure {
         this.collectedLocalLinks = new ArrayList<>();
         this.globalAnchorsById = new HashMap<>();
         this.localAnchorIdsByTocItem = new HashMap<>();
+        this.uniqueAnchorIdGenerator = new UniqueAnchorIdGenerator();
     }
 
     void removeGlobalAnchorsForPath(Path path) {
@@ -118,6 +120,16 @@ class WebSiteDocStructure implements DocStructure {
     @Override
     public String fullUrl(String relativeUrl) {
         return  "/" + docMeta.getId() + "/" + relativeUrl;
+    }
+
+    @Override
+    public void onSectionOrSubHeading(Path path, int level, String id) {
+        uniqueAnchorIdGenerator.registerSectionOrSubHeading(path, level, id);
+    }
+
+    @Override
+    public String generateUniqueAnchor(Path path, String localId) {
+        return uniqueAnchorIdGenerator.generateId(path, localId);
     }
 
     @Override
@@ -212,9 +224,6 @@ class WebSiteDocStructure implements DocStructure {
 
     private Optional<String> validateExternalLink(LinkToValidate linkToValidate) {
         String url = linkToValidate.docUrl.getUrl();
-//        if (url.endsWith("/")) {
-//            url += "index.html";
-//        }
 
         int responseCode = pingUrlConnection(url, 5000);
         if ((responseCode >= 200 && responseCode <= 204) || responseCode == 301) {

--- a/znai-website-gen/src/test/groovy/org/testingisdocumenting/znai/extensions/table/TableFencePluginTest.groovy
+++ b/znai-website-gen/src/test/groovy/org/testingisdocumenting/znai/extensions/table/TableFencePluginTest.groovy
@@ -40,7 +40,7 @@ Account, Price, Description
                                          [[markup: "chair", type: "TestMarkup"]]],
                                         [[[markup: "#91AGB1", type: "TestMarkup"]],
                                          [[markup: 10, type: "TestMarkup"]],
-                                         [[markup: "lunch", type: "TestMarkup"]]]]], title: "my table"]
+                                         [[markup: "lunch", type: "TestMarkup"]]]]], title: "my table", anchorId: "my-table"]
 
     }
 

